activation/malfeasance2.go

```diff
@@ -3,16 +3,96 @@ package activation
 import (
 	"context"
 	"fmt"
+	"sync"
+
+	"go.uber.org/zap"
 
 	"github.com/spacemeshos/go-spacemesh/activation/wire"
 	"github.com/spacemeshos/go-spacemesh/codec"
 	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/signing"
 )
 
-type MalfeasanceHandlerV2 struct{}
+type MalfeasanceHandlerV2 struct {
+	logger *zap.Logger
+
+	malPublisher malfeasancePublisher
+	edVerifier   *signing.EdVerifier
+	validator    nipostValidatorV2
+
+	smeshingMutex sync.Mutex
+	signers       map[types.NodeID]*signing.EdSigner
+}
+
+func NewMalfeasanceHandlerV2(
+	logger *zap.Logger,
+	malPublisher malfeasancePublisher,
+	edVerifier *signing.EdVerifier,
+	validator nipostValidatorV2,
+) *MalfeasanceHandlerV2 {
+	return &MalfeasanceHandlerV2{
+		logger:       logger,
+		malPublisher: malPublisher,
+		edVerifier:   edVerifier,
+		validator:    validator,
+
+		signers: make(map[types.NodeID]*signing.EdSigner),
+	}
+}
+
+func (p *MalfeasanceHandlerV2) Register(sig *signing.EdSigner) {
+	p.smeshingMutex.Lock()
+	defer p.smeshingMutex.Unlock()
+	if _, exists := p.signers[sig.NodeID()]; exists {
+		p.logger.Error("signing key already registered", log.ZShortStringer("id", sig.NodeID()))
+		return
+	}
 
-func NewMalfeasanceHandlerV2() *MalfeasanceHandlerV2 {
-	return &MalfeasanceHandlerV2{}
+	p.logger.Debug("registered signing key", log.ZShortStringer("id", sig.NodeID()))
+	p.signers[sig.NodeID()] = sig
+}
+
+// Publish publishes an ATX proof by encoding it and sending it to the malfeasance publisher.
+func (p *MalfeasanceHandlerV2) Publish(ctx context.Context, nodeID types.NodeID, proof wire.Proof) error {
+	p.smeshingMutex.Lock()
+	_, exists := p.signers[nodeID]
+	p.smeshingMutex.Unlock()
+
+	if exists {
+		// do not publish proofs against one self
+		return fmt.Errorf("publish ATX malfeasance proof: identity %s is managed by node", nodeID)
+	}
+
+	proofNodeID, err := proof.Valid(ctx, p)
+	if err != nil {
+		return fmt.Errorf("publish ATX malfeasance proof: proof not valid: %w", err)
+	}
+	if proofNodeID != nodeID {
+		return fmt.Errorf("publish ATX malfeasance proof: proof for %s does not match node ID %s",
+			proofNodeID.ShortString(), nodeID.ShortString(),
+		)
+	}
+
+	atxProof := &wire.ATXProof{
+		Version:   0x01, // for now we only have one version
```

Comment:
Makes sense to have a named constant here

Reply:
I added this, but again I think this just eagerly adds more to the codebase that might never be needed: https://github.com/spacemeshos/go-spacemesh/pull/6307/commits/8a365a57228617a0a0048e50c0b1416debc96d8d 😅